api-fput-object.go

```diff
@@ -0,0 +1,204 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2015 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package minio
+
+import (
+	"os"
+	"sort"
+)
+
+// getUploadID if already present for object name or initiate a request to fetch a new upload id.
+func (c Client) getUploadID(bucketName, objectName, contentType string) (string, error) {
+	// Input validation.
+	if err := isValidBucketName(bucketName); err != nil {
+		return "", err
+	}
+	if err := isValidObjectName(objectName); err != nil {
+		return "", err
+	}
+
+	// Set content Type to default if empty string.
+	if contentType == "" {
+		contentType = "application/octet-stream"
+	}
+
+	// Find upload id for previous upload for an object.
+	uploadID, err := c.findUploadID(bucketName, objectName)
+	if err != nil {
+		return "", err
+	}
+	if uploadID == "" {
+		// Initiate multipart upload for an object.
+		initMultipartUploadResult, err := c.initiateMultipartUpload(bucketName, objectName, contentType)
+		if err != nil {
+			return "", err
+		}
+		// Save the new upload id.
+		uploadID = initMultipartUploadResult.UploadID
+	}
+	return uploadID, nil
+}
+
+// getMultipartStat gather next part number, total uploaded size and list of parts.
+// TODO: reduce number of results.
+func (c Client) getMultipartStat(bucketName, objectName, uploadID string) (int, int64, completeMultipartUpload, error) {
+	// Input validation.
```

Comment:
returning 4 variables is fine, but in cases like these int and int64 does not convey what it stands for. Either we can document them or declare variables here (preferred approach).